### PR TITLE
[Core][Sleep Mode] Add opt-in preparation of first-sleep backup buffers

### PR DESCRIPTION
**Pull request by AlanFokCo**
## Purpose

The first level-1 sleep can spend most of its time allocating pinned host buffers for weight backup. Applications that put the first sleep on a latency-sensitive path may prefer to pay that allocation cost during startup.

This adds `VLLM_SLEEP_PREPARE_BACKUP_MAX_BYTES` (default `0`, disabled). With CUDA and the CuMem sleep backend enabled, the worker prepares empty weight-backup buffers after warmup, before becoming ready. The first sleep still copies the **current** GPU weights, including updates made after preparation. Prepared capacity is consumed once; later sleeps use the existing allocator/cache behavior.

The budget covers the complete set of prepared weight buffers, conservatively rounding each allocation up to a power of two. Preparation is skipped if that set exceeds the budget. Allocation changes, reload, discard, level-2 sleep and shutdown cancel unused preparation. The change is limited to the allocator, worker lifecycle, configuration, documentation and tests; it introduces no new public RPC.

This trades earlier host-memory residency and longer startup for a shorter first sleep. It does not remove the allocation cost or speed up weight copying. The budget is per worker, **not** a process-wide or machine-wide pinned-memory limit; existing host cache and other allocations are outside it. Cancellation drops references without flushing the global host cache.

Error handling is deliberately conservative: `torch.OutOfMemoryError` and `MemoryError` cancel preparation and fall back to ordinary sleep; other exceptions release prepared references and propagate. In the tested PyTorch version, CUDA pinned-allocation exhaustion can surface as `torch.AcceleratorError` and fail startup. This option does not guarantee recovery from every allocation failure or an OS out-of-memory kill.

Related work checked: #55206 releases host cache after wake, and #57629 changes wake remapping order for unified memory. This PR instead prepares capacity before the first sleep. It neither implements nor depends on the graph offload/discard mechanisms in #56373 or #57899. No open PR implementing the same startup preparation was found in the duplicate-work search. If another policy clears the host cache later, this change does not replenish prepared buffers or avoid the next allocation cost.

## Test Plan

```bash
python -m pytest tests/v1/worker/test_sleep_mode_backend.py -q
python -m pytest tests/basic_correctness/test_mem.py -q -s \
  -k 'prepared_sleep_backups or basic_cumem'
```

The unit tests cover disabled preparation, rounded budgets, partial allocation failure, unexpected error propagation, stale allocation identity/handles, cancellation and worker lifecycle hooks. The GPU test changes weights **after** preparation, disallows new CPU backup allocation during the first sleep, verifies the updated values after wake, and then exercises another sleep/wake cycle. Injected allocation exceptions do not emulate real host-memory exhaustion.

For the model experiments, both the unchanged baseline and this candidate used one H20, FP16, TP=1, the V2 runner, no `torch.compile`, FULL graph capture sizes `[1, 2, 4, 8]`, synchronous scheduling, `max_model_len=256`, `max_num_seqs=8`, `max_num_batched_tokens=256`, prefix caching disabled, and `gpu_memory_utilization=0.8`. The candidate used `VLLM_SLEEP_PREPARE_BACKUP_MAX_BYTES=137438953472` (128 GB, with 1 GB = 2^30 bytes); the baseline used `0`. This is the configured budget, not measured physical pinned residency.

Each model/configuration had three independent startups and 20 level-1 sleep/wake cycles per startup. A separate local benchmark recorded startup, first sleep, and subsequent sleep/wake times without internal timing instrumentation. After every wake, the first generation and five subsequent generations were compared with the same process's initial token IDs: four fixed prompts, temperature 0, seed 42, `ignore_eos=True`, 32 generated tokens per prompt. These are controlled output-regression checks, not a broad accuracy evaluation. The benchmark script is separate from the tests included in this diff.

## Test Result

Local H20 validation: driver 580.126.09, Python 3.12.3, PyTorch 2.13.0+cu130. Baseline: `e9f169d16b`. The native wheel was built from `97dc6b19d2`; a source comparison found no differences between these revisions in `csrc`, `cmake`, `CMakeLists.txt`, `requirements`, `setup.py`, `vllm/_custom_ops.py`, or `vllm/third_party`. These results are local validation, not upstream CI results.

- Unit tests: **29 passed**.
- CUDA allocator tests: **2 passed, 19 deselected**.
- Model comparisons: **18 independent startups, 360 cycles** across baseline and candidate; all controlled output checks passed.
- Targeted pre-commit checks passed.

Times below are seconds. Each cell is the median across three independent startups. For subsequent sleeps, each startup contributes the median of cycles 2–20. “Startup + first sleep” is summed per run before taking the median.

| Model (FP16, TP=1) | First sleep, baseline → prepared | Startup + first sleep, baseline → prepared | Subsequent sleep, baseline → prepared |
|---|---:|---:|---:|
| Qwen3-0.6B | 0.618 → 0.059 | 16.209 → 16.275 | 0.057 → 0.057 |
| Qwen3-8B | 6.193 → 0.343 | 23.905 → 23.747 | 0.343 → 0.343 |
| Qwen3-30B-A3B-Instruct-2507 | 21.899 → 1.160 | 48.201 → 48.150 | 1.159 → 1.159 |

The data supports moving the first allocation cost out of sleep; it does not demonstrate a comparable reduction in total startup-plus-sleep time or a subsequent-cycle speedup. Multi-GPU, unified-memory systems and real host-exhaustion recovery were not validated. Some model runs on both trees emitted multiprocessing semaphore-cleanup warnings at exit; they returned exit code 0 and are not described as warning-free runs.

AI assistance was used for implementation, tests, local experiment execution and review. The author is responsible for the submitted change; these results are not a claim that every test was personally executed by the author.

---
<details>
<summary>Essential elements</summary>

- [x] Purpose and relationship to existing work described.
- [x] Test commands and validation scope provided.
- [x] Before/after results and limitations reported.
- [x] Sleep-mode documentation updated.

</details>

### CODE REVIEW

**Comment by AlanFokCo** [comment]
A little more context on the intended benefit: this option moves pinned-host buffer allocation from the first level-1 sleep into startup, so an already-started engine can give up its GPU memory sooner when asked to sleep. The weight copy still happens at sleep time, using the current weights.

On one H20 (FP16, TP=1), the median first-sleep times across three independent startups were:

| Model | Baseline | With preparation |
|---|---:|---:|
| Qwen3-0.6B | 0.618 s | 0.059 s |
| Qwen3-8B | 6.193 s | 0.343 s |
| Qwen3-30B-A3B-Instruct-2507 | 21.899 s | 1.160 s |

The intended use case is a deployment that can schedule startup ahead of time but needs a short first resource handoff: for example, an inference engine sharing a GPU with another workload, or the first inference-to-training transition. These are potential applications of the measured first-sleep improvement; the tests here were sleep/wake experiments, not an end-to-end scheduler or training benchmark.

The tradeoff matters: startup takes longer and pinned host memory becomes resident earlier. For the 30B model, startup plus first sleep remained approximately 48.2 s with either configuration. Subsequent sleep times were essentially unchanged because the existing host allocator can already reuse its cache. This does not reduce steady GPU residency, speed up weight copying, or avoid reallocation if later graph capture clears the host cache. The option is disabled by default, and its per-worker preparation budget is not a process-wide physical-memory limit.

@aoshen02 Would you be willing to review this proposal, particularly whether this startup-versus-first-sleep tradeoff is useful for vLLM's sleep/wake users and whether the configuration and allocator lifecycle are appropriate? The PR description includes the validation scope and allocation-failure limitations. This change is independent of the graph offload/discard work.
